vkcs/compute/resource_instance.go

```diff
@@ -1390,41 +1399,69 @@ func resourceComputeSchedulerHintsHash(v interface{}) int {
 	return hashcode.String(buf.String())
 }
 
-func checkBlockDeviceConfig(d *schema.ResourceData) error {
-	if vL, ok := d.GetOk("block_device"); ok {
-		vLs := vL.([]interface{})
-		isBootIdxZeroSet := (len(vLs) <= 1)
+func checkBlockDeviceConfig(d *schema.ResourceData) diag.Diagnostics {
+	vL, ok := d.GetOk("block_device")
+	if !ok {
+		return nil
+	}
 
-		for _, v := range vLs {
-			vM := v.(map[string]interface{})
+	vLs := vL.([]interface{})
+	isBootIdxZeroSet := len(vLs) <= 1
+	for _, v := range vLs {
+		vM := v.(map[string]interface{})
 
-			if vM["source_type"] != "blank" && vM["uuid"] == "" {
-				return fmt.Errorf("you must specify a uuid for %s block device types", vM["source_type"])
-			}
+		if vM["source_type"] != "blank" && vM["uuid"] == "" {
+			return diag.Errorf("you must specify a uuid for %s block device types", vM["source_type"])
+		}
 
-			if vM["source_type"] == "image" && vM["destination_type"] == "volume" {
-				if vM["volume_size"] == 0 {
-					return fmt.Errorf("you must specify a volume_size when creating a volume from an image")
-				}
+		if vM["source_type"] == "image" && vM["destination_type"] == "volume" {
+			if vM["volume_size"] == 0 {
+				return diag.Errorf("you must specify a volume_size when creating a volume from an image")
 			}
+		}
 
-			if vM["source_type"] == "blank" && vM["destination_type"] == "local" {
-				if vM["volume_size"] == 0 {
-					return fmt.Errorf("you must specify a volume_size when creating a blank block device")
-				}
+		if vM["source_type"] == "blank" && vM["destination_type"] == "local" {
+			if vM["volume_size"] == 0 {
+				return diag.Errorf("you must specify a volume_size when creating a blank block device")
 			}
+		}
 
-			if vM["boot_index"] == 0 {
-				isBootIdxZeroSet = true
-			}
+		if vM["boot_index"] == 0 {
+			isBootIdxZeroSet = true
 		}
+	}
 
-		if !isBootIdxZeroSet {
-			return fmt.Errorf("you must set boot_index to 0 for one of block_devices")
+	if !isBootIdxZeroSet {
+		return diag.Errorf("you must set boot_index to 0 for one of block_devices")
+	}
+
+	diags := diag.Diagnostics{}
+	for _, v := range vLs {
+		vM := v.(map[string]interface{})
+		deleteOnTermination := vM["delete_on_termination"].(bool)
+		sourceType := vM["source_type"].(string)
+		switch {
+		case sourceType == "blank" || sourceType == "image" || sourceType == "snapshot":
+			if !deleteOnTermination {
+				diags = append(diags, diag.Diagnostic{
+					Severity: diag.Warning,
+					Summary: fmt.Sprintf("block_device %s: delete_on_termination should be true, when source_type is %s",
+						vM["uuid"].(string), sourceType),
+				})
+				break
+			}
+		case sourceType == "volume":
+			if deleteOnTermination {
+				diags = append(diags, diag.Diagnostic{
+					Severity: diag.Warning,
+					Summary:  fmt.Sprintf("block_device %s: delete_on_termination should be false, when source_type is volume", vM["uuid"].(string)),
+				})
+				break
```

Comment:
In both cases above you can return right away, as slice of diags will always be of length one. Also, it would be nice to add utility `Warningf` similar to `diag.Errorf`, you can place it in internal/util.

Reply:
done